### ISSUE TITLE
Decimal multiplication and division overflow for no reason 

### ISSUE THREAD

**Issue by arthurvinci** [radixdlt/radixdlt-scrypto#585]
The functions `Div` and `Mul` overflow for no reason. The following pieces of code return overflow errors:
```Rust
let test = Decimal::MAX*1;
```
```Rust
let test = Decimal::MAX/1;
```

For the Mul function, the error comes from the following piece of code in `decimal.rs`:
```Rust
let a = self.0;
let b_dec: Decimal = other.try_into().expect("Overflow");
let b: I256 = b_dec.0;
let c: I256 = a * b / Self::ONE.0;
```
If we call `1_b` the binary number representing `1` in `Decimal`
Then `a*b` can be in [2^256 + 1_b -1; 0]  but should still not throw an overflow. Therefore, the multiplication function will overflow for all `a`, `b` such that `a*b` is in [2^256 +1_b - 1; 2^256 -1].
Note that changing the line `c = a * b / Self::ONE.0;` to `c = a / Self::ONE.0 * b` creates an issue. If `a` is smaller than `1` then the multiplication will always return `0`. 

For the Division function, the error comes from the following piece of code in `decimal.rs`:
```Rust
let a = self.0;
let b_dec: Decimal = other.try_into().expect("Overflow");
let b: I256 = b_dec.0;
let c: I256 = a * Self::ONE.0 / b;
```
We have a similar issue.
`a*Self::ONE.0` can be in [2^256 + c -1; 0] but the final division `a/b` not overflow in theory.
Note: Changing the order of the operations in the last line also creates an issue. If `a<b` then `a/b`will always return 0.

**Comment by arthurvinci** [comment]
Proposed solutions:
For multiplication:
```Rust
    fn mul(self, other: T) -> Self::Output {
        let a: I512 = I512::from(self.0);
        let b_dec: Decimal = other.try_into().expect("Overflow");
        let b: I512 = I512::from(b_dec.0);
        let c: I512 = a * b / I512::from(Self::ONE.0);
        let c_256 = I256::try_from(c).expect("Overflow");
        Decimal(c_256)
    }
```
For division:
```Rust
fn div(self, other: T) -> Self::Output {
        let a: I512 = I512::from(self.0);
        let b_dec: Decimal = other.try_into().expect("Overflow");
        let b: I512 = I512::from(b_dec.0);
        let c: I512 = a * I512::from(Self::ONE.0) / b;
        let c_256 = I256::try_from(c).expect("Overflow");
        Decimal(c_256)
    }
```    
I could not test the code because of the issue (#584). 

A better solution would consist in having a `IX` representation with `X = 256 + log_2(10^18)` and casting in this interface to avoid using `I512` representation which is too big.

**Comment by arthurvinci** [comment]
Solved by #604
